patroni/ha.py

```diff
@@ -46,6 +46,50 @@ def failover_limitation(self):
         return None
 
 
+class BackgroundKeepaliveSender(object):
+    """A context manager that sends keepalives every loop_wait seconds in a background thread while the context is
+    running, but only after a safepoint has been reached. After the safepoint PostgreSQL must not be allowed to
```

Comment:
And to answer your questions we don't want to send keepalives before safepoint because shutting down PostgreSQL might not work correctly. Keepalive can be sent when:

1. We have updated our leader lock.
2. PostgreSQL can't send back successful commits before further action is taken - either it's in standby mode, postgres is shut down or postgres is in process of shutting down and all backends have been closed.